### PR TITLE
Orchestration spi design in Sharding-proxy

### PR DESCRIPTION
**Pull request by yu199195**

### CODE REVIEW

File: shardingsphere-proxy/shardingsphere-proxy-common/src/main/java/org/apache/shardingsphere/proxy/config/converter/ProxyConfigurationConverterFactory.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    
+    /**
+     * New proxy configuration converter instances.
+     *
+     * @param isOrchestration is orchestration
+     * @return proxy configuration converter
+     */
+    public static ProxyConfigurationConverter newInstances(final boolean isOrchestration) {
+        if (isOrchestration) {
+            Optional<ProxyConfigurationConverter> configurationConverter = SingletonServiceLoader.getServiceLoader(ProxyConfigurationConverter.class).newServiceInstances();
```
Comment (kimmking) [inline shardingsphere-proxy/shardingsphere-proxy-common/src/main/java/org/apache/shardingsphere/proxy/config/converter/ProxyConfigurationConverterFactory.java:37]: extract a new method

File: shardingsphere-proxy/shardingsphere-proxy-common/src/main/java/org/apache/shardingsphere/proxy/config/converter/DefaultConfigurationConverter.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+import org.apache.shardingsphere.kernel.context.SchemaContextsAware;
+import org.apache.shardingsphere.kernel.context.SchemaContextsBuilder;
+import org.apache.shardingsphere.kernel.context.schema.DataSourceParameter;
+import org.apache.shardingsphere.proxy.config.ProxyConfiguration;
+import org.apache.shardingsphere.proxy.config.ShardingConfiguration;
+
+/**
+ * Default configuration converter.
+ */
+public class DefaultConfigurationConverter extends AbstractConfigurationConverter {
```
Comment (kimmking) [inline shardingsphere-proxy/shardingsphere-proxy-common/src/main/java/org/apache/shardingsphere/proxy/config/converter/DefaultConfigurationConverter.java:39]: final

File: shardingsphere-proxy/shardingsphere-proxy-frontend/shardingsphere-proxy-frontend-mysql/src/test/java/org/apache/shardingsphere/proxy/frontend/mysql/MySQLProtocolFrontendEngineTest.java [hunk 3]
```diff
@@ -171,4 +142,26 @@ private void setConnectionPhase(final MySQLConnectionPhase connectionPhase) {
         field.setAccessible(true);
         field.set(mysqlProtocolFrontendEngine.getAuthEngine(), connectionPhase);
     }
+    
+    @SneakyThrows
+    private void initProxySchemaContexts(final Authentication authentication) {
+        Field field = ProxySchemaContexts.getInstance().getClass().getDeclaredField("schemaContexts");
```
Comment (kimmking) [inline shardingsphere-proxy/shardingsphere-proxy-frontend/shardingsphere-proxy-frontend-mysql/src/test/java/org/apache/shardingsphere/proxy/frontend/mysql/MySQLProtocolFrontendEngineTest.java:148]: why not add a setter?

File: shardingsphere-proxy/shardingsphere-proxy-frontend/shardingsphere-proxy-frontend-mysql/src/test/java/org/apache/shardingsphere/proxy/frontend/mysql/MySQLProtocolFrontendEngineTest.java [hunk 4]
```diff
@@ -171,4 +142,26 @@ private void setConnectionPhase(final MySQLConnectionPhase connectionPhase) {
...
+        field.setAccessible(true);
+        field.set(ProxySchemaContexts.getInstance(), getSchemaContexts(authentication));
+    }
+    
+    private SchemaContexts getSchemaContexts(final Authentication authentication) {
+        return new SchemaContexts(getSchemaContextMap(), new ConfigurationProperties(new Properties()), authentication);
+    }
+    
+    private Map<String, SchemaContext> getSchemaContextMap() {
+        Map<String, SchemaContext> result = new HashMap<>(10);
```
Comment (kimmking) [inline shardingsphere-proxy/shardingsphere-proxy-frontend/shardingsphere-proxy-frontend-mysql/src/test/java/org/apache/shardingsphere/proxy/frontend/mysql/MySQLProtocolFrontendEngineTest.java:158]: remove 10, default is 16

File: shardingsphere-proxy/shardingsphere-proxy-frontend/shardingsphere-proxy-frontend-mysql/src/test/java/org/apache/shardingsphere/proxy/frontend/mysql/MySQLProtocolFrontendEngineTest.java [hunk 5]
```diff
@@ -171,4 +142,26 @@ private void setConnectionPhase(final MySQLConnectionPhase connectionPhase) {
...
+        for (int i = 0; i < 10; i++) {
+            String name = "schema_" + i;
```
Comment (kimmking) [inline shardingsphere-proxy/shardingsphere-proxy-frontend/shardingsphere-proxy-frontend-mysql/src/test/java/org/apache/shardingsphere/proxy/frontend/mysql/MySQLProtocolFrontendEngineTest.java:160]: constants

File: shardingsphere-proxy/shardingsphere-proxy-common/src/main/java/org/apache/shardingsphere/proxy/config/converter/DefaultConfigurationConverter.java (hunk 2 above)
Comment (yu199195) [reply shardingsphere-proxy/shardingsphere-proxy-common/src/main/java/org/apache/shardingsphere/proxy/config/converter/DefaultConfigurationConverter.java:39]: ok

File: shardingsphere-proxy/shardingsphere-proxy-frontend/shardingsphere-proxy-frontend-mysql/src/test/java/org/apache/shardingsphere/proxy/frontend/mysql/MySQLProtocolFrontendEngineTest.java (hunk 3 above)
Comment (yu199195) [reply shardingsphere-proxy/shardingsphere-proxy-frontend/shardingsphere-proxy-frontend-mysql/src/test/java/org/apache/shardingsphere/proxy/frontend/mysql/MySQLProtocolFrontendEngineTest.java:148]: is areadly exist in before， mybe use Field more reasonable

File: shardingsphere-proxy/shardingsphere-proxy-frontend/shardingsphere-proxy-frontend-mysql/src/test/java/org/apache/shardingsphere/proxy/frontend/mysql/MySQLProtocolFrontendEngineTest.java (hunk 5 above)
Comment (yu199195) [reply shardingsphere-proxy/shardingsphere-proxy-frontend/shardingsphere-proxy-frontend-mysql/src/test/java/org/apache/shardingsphere/proxy/frontend/mysql/MySQLProtocolFrontendEngineTest.java:160]: ok

File: shardingsphere-proxy/shardingsphere-proxy-common/src/main/java/org/apache/shardingsphere/proxy/config/converter/AbstractConfigurationConverter.java [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+     * @param localRuleConfigs rule configs for YAML.
+     * @return rule configurations
+     */
+    protected Map<String, Collection<RuleConfiguration>> getRuleConfigurations(final Map<String, YamlProxyRuleConfiguration> localRuleConfigs) {
+        YamlRuleConfigurationSwapperEngine swapperEngine = new YamlRuleConfigurationSwapperEngine();
+        return localRuleConfigs.entrySet().stream().collect(Collectors.toMap(Map.Entry::getKey, entry -> swapperEngine.swapToRuleConfigurations(entry.getValue().getRules())));
+    }
+    
+    /**
+     * Gets data source parameters map.
```
Comment (menghaoranss) [inline shardingsphere-proxy/shardingsphere-proxy-common/src/main/java/org/apache/shardingsphere/proxy/config/converter/AbstractConfigurationConverter.java:51]: May be `Get`

File: shardingsphere-proxy/shardingsphere-proxy-common/src/main/java/org/apache/shardingsphere/proxy/config/converter/AbstractConfigurationConverter.java [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+     *
+     * @param localRuleConfigs rule configs for YAML.
+     * @return data source parameters map
+     */
+    protected Map<String, Map<String, DataSourceParameter>> getDataSourceParametersMap(final Map<String, YamlProxyRuleConfiguration> localRuleConfigs) {
+        return localRuleConfigs.entrySet().stream().collect(Collectors.toMap(Map.Entry::getKey, entry -> getDataSourceParameters(entry.getValue().getDataSources())));
+    }
+    
+    /**
+     * Gets data source parameters.
```
Comment (menghaoranss) [inline shardingsphere-proxy/shardingsphere-proxy-common/src/main/java/org/apache/shardingsphere/proxy/config/converter/AbstractConfigurationConverter.java:61]: Same as above.

File: shardingsphere-proxy/shardingsphere-proxy-common/src/main/java/org/apache/shardingsphere/proxy/config/converter/AbstractConfigurationConverter.java [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+/**
+ * Abstract configuration converter.
+ */
+public abstract class AbstractConfigurationConverter implements ProxyConfigurationConverter {
+    
+    /**
+     * Get rule configurations.
+     *
+     * @param localRuleConfigs rule configs for YAML.
```
Comment (menghaoranss) [inline shardingsphere-proxy/shardingsphere-proxy-common/src/main/java/org/apache/shardingsphere/proxy/config/converter/AbstractConfigurationConverter.java:42]: `param` comment should not end with '.'

File: shardingsphere-proxy/shardingsphere-proxy-common/src/main/java/org/apache/shardingsphere/proxy/config/converter/AbstractConfigurationConverter.java (hunk 7 above)
Comment (yu199195) [reply shardingsphere-proxy/shardingsphere-proxy-common/src/main/java/org/apache/shardingsphere/proxy/config/converter/AbstractConfigurationConverter.java:61]: ok

File: shardingsphere-proxy/shardingsphere-proxy-common/src/main/java/org/apache/shardingsphere/proxy/config/converter/AbstractConfigurationConverter.java (hunk 6 above)
Comment (yu199195) [reply shardingsphere-proxy/shardingsphere-proxy-common/src/main/java/org/apache/shardingsphere/proxy/config/converter/AbstractConfigurationConverter.java:51]: ok

File: shardingsphere-proxy/shardingsphere-proxy-common/src/main/java/org/apache/shardingsphere/proxy/config/converter/AbstractConfigurationConverter.java (hunk 8 above)
Comment (yu199195) [reply shardingsphere-proxy/shardingsphere-proxy-common/src/main/java/org/apache/shardingsphere/proxy/config/converter/AbstractConfigurationConverter.java]: ok，will fix

**Pull request merged**